Moving fast parts of cast FCALLs to managed code.

VSadov opened the pull request:
* Exposed casting cache to managed code
* Implemented a managed version of cache lookup
* Moved `JIT_IsInstanceOfAny` and `JIT_ChkCastAny`  to managed code as the first ones to move.
* Skip managed JIT helpers in exception stack traces and debugger
* Managed `JIT_IsInstanceOfInterface`
* All other cast helpers are managed.

Fixes:https://github.com/dotnet/coreclr/issues/27931

VSadov committed:
Make the CastCache table rooted by a static field.

VSadov committed:
Moving fast parts of casting FCALLs to managed code

VSadov committed:
Implemented managed cache lookup

andrew-boyarshin commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs:
```diff
@@ -0,0 +1,222 @@
+        }
+
+        // NOTE!!
+        // This is a copy of C++ implementation in CastCache.cpp
```
More likely `castcache.h`.

jkotas replied:
Nit: The implementation is in the .cpp file, but the VM filename is in lower case (castcache.cpp).

VSadov commented:
@tommcdon - tried the "logical unwind" here in the last change. Does not seem to be sufficient. 
VS still visualizes invalid casts as rethrows.

tommcdon commented:
> @tommcdon - tried the "logical unwind" here in the last change. Does not seem to be sufficient.
> VS still visualizes invalid casts as rethrows.

Proposed fix at a lower level is here: https://github.com/VSadov/runtime/commit/d269b628cdcd0c61257a4b15a53c2ff074efaa04. Tested the fix in VS and it seems to work correctly.

VSadov committed:
Simple hiding from traces/debugger

VSadov committed:
Managed JIT_IsInstanceOfInterface

VSadov committed:
deleted the array casting helpers. not special enough now

VSadov committed:
reconcilled changes between managed and native versions of TryGet

VSadov committed:
managed JIT_IsInstanceOfClass

VSadov committed:
remaining casting helpeers are managed

VSadov committed:
move helpers init to ECall.cpp

VSadov committed:
some perf tweaks

VSadov committed:
Cleanup for review

VSadov committed:
Add `AggressiveOptimization`

VSadov commented:
I think this is ready to be reviewed.

VSadov commented:
The change moves fast parts of casting helpers to managed code and removes various existing implementations - in C++, in asm in C/inlined asm. Now it is all C#.

VSadov commented:
I also hoped to not have specialized versions for class and interface casts. 

Tuns out outperforming hand-tuned assembly is challenging, so specialized methods are still there.
I think cache lookup must get 20-30% faster before considering just using cache. As it is, best case scenarios for classes and interfaces are noticeably faster than cache lookuos and these scenarios also seem fairly common.

Whoever wrote the AMD64 assembly helpers did a god job.

I have discussed with @davidwrighton possible ways of speeding up the cache lookup:
- use singleton table instead of `null` for cache flush (will eliminate one null check)
- use fixed size table (will simplify reducing hashcode to table index)
- consider simpler hash function, perhaps at cost of larger table.
- inline cache lookup by hand (some codepaths could be streamlined)
- precompute hashed value of methodtable pointer.
. . . something else..

This may improve the cache lookup or maybe it will not.

I think perf tinkering should be a separate change though. 
This PR achieves its goals - the helpers are managed and perf is roughly the same or better.

jkotas commented on src/coreclr/src/inc/corinfo.h:
```diff
@@ -423,12 +423,10 @@ enum CorInfoHelpFunc
     // the right helper to use
 
     CORINFO_HELP_ISINSTANCEOFINTERFACE, // Optimized helper for interfaces
-    CORINFO_HELP_ISINSTANCEOFARRAY,  // Optimized helper for arrays
```
I would keep these for now, even if it means that the implementation is same.

VSadov replied:
I can get them back.

---
In reply to: [368260447](https://github.com/dotnet/runtime/pull/1068#discussion_r368260447) [](ancestors = 368260447)

VSadov replied:
I just could not see what could be done differently in a helper if we knew we care casting to an array. 
For identity cast check we could just compare MTs, which we do for any cast...

---
In reply to: [368261248](https://github.com/dotnet/runtime/pull/1068#discussion_r368261248) [](ancestors = 368261248,368260447)

VSadov replied:
I am guessing JIT could inline identity cast check similarly to class cast. Not sure if those are common with arrays but  with classes inlining identity cast helps a lot.

---
In reply to: [368261350](https://github.com/dotnet/runtime/pull/1068#discussion_r368261350) [](ancestors = 368261350,368261248,368260447)

VSadov commented:
Here is what I see on a microbenchmark that calls a method that performs a particular cast in a loop 

```

Empty: 232ms      <-- overhead, should be _roughly_ subtracted from other measurements.
Base to Base: 417ms    <-- inlined by the JIT, nearly free
Derived to Base: 504ms    <-- IsInstanceOfClassSpecial  1 iteration
Derived2 to Base: 552ms   <-- IsInstanceOfClassSpecial  3 iterations
Derived4 to Base: 622ms   <-- IsInstanceOfClassSpecial  5 iterations
Derived6 to Base: 754ms   <-- IsInstanceOfClassSpecial  7 iterations. no-nullcheck IsInstanceAny would have similar cost. 

List<int> to IList<int>: 601ms       <-- IsInstanceOfInterface 1 iteration
List<int> to ICollection<int>: 644ms <-- IsInstanceOfInterface 2 iterations
List<int> to IEnumerable<int>: 861ms <-- IsInstanceAny   (3rd, but JIT does not like variant interfaces)
List<int> to ICollection: 589ms      <-- IsInstanceOfInterface 5 iterations
List<double> to ICollection<int>: 719ms   <-- IsInstanceOfInterface 8 iterations
Dictionary<double, double> to IDeserializationCallback: 859ms   <-- IsInstanceOfInterface 10 iterations.

List<int> to IReadOnlyCollection<int>: 1062ms  <-- IsInstanceAny, looks like a hash collision
List<string> to IReadOnlyCollection<object>: 842ms  <-- IsInstanceAny  pass
List<double> to IReadOnlyCollection<int>: 966ms     <-- IsInstanceAny  failure 
string[] to IReadOnlyCollection<object>: 858ms      <-- IsInstanceAny  pass
```

where the baseline is 

```
Empty: 352ms	<-- overhead, no idea why slower than Base to Base. branch misprediction?
Base to Base: 325ms
Derived to Base: 627ms
Derived2 to Base: 599ms
Derived4 to Base: 691ms
Derived6 to Base: 789ms

List<int> to IList<int>: 550ms
List<int> to ICollection<int>: 724ms
List<int> to IEnumerable<int>: 1075ms  <-- new IsInstanceAny is a bit faster (due to cache improvements, I think).
List<int> to ICollection: 739ms
List<double> to ICollection<int>: 783ms
Dictionary<double, double> to IDeserializationCallback: 919ms

List<int> to IReadOnlyCollection<int>: 973ms
List<string> to IReadOnlyCollection<object>: 1365ms
List<double> to IReadOnlyCollection<int>: 1015ms
string[] to IReadOnlyCollection<object>: 971ms
```

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:
```diff
@@ -343,5 +366,37 @@ public int MultiDimensionalArrayRank
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                int mtFieldOffset = 0x10 / sizeof(nuint);
```
Why this cannot be just a regular field access? The ifdefs can be on the field offset.

VSadov replied:
There would be 3 combinations for BaseMethodTable and 4 combinations for the InterfaceMap (32/64bit X Ret/Debug)

---
In reply to: [368260513](https://github.com/dotnet/runtime/pull/1068#discussion_r368260513) [](ancestors = 368260513)

jkotas replied:
Use `const` + ifdefs to make it look pretty?

VSadov replied:
I will try, but I think it might look worse than this.

---
In reply to: [368261592](https://github.com/dotnet/runtime/pull/1068#discussion_r368261592) [](ancestors = 368261592,368260513)

jkotas replied:
Use `const` + ifdefs to make it look pretty?

VSadov replied:
that is the idea. These can be constant expressions, I think.

---
In reply to: [368262114](https://github.com/dotnet/runtime/pull/1068#discussion_r368262114) [](ancestors = 368262114)

VSadov commented:
The Linux is similar, but there are few interesting differences.

Since we did not use hand-written assembly on Linux and C++ helpers were not as sophisticated, there is some gain in specialized methods.

On the other hand the improvements in cache lookup seems to have been negated by something. We even see a 1-2% regression. I think it is acceptable, but wonder why we have it.

Either we have some platform specific penalty for moving to managed (call conventions, access of statics?),  or GCC was producing better code for native cache lookup and JIT is not matching that.
It would be interesting to find out. Maybe we can improve something.

===== baseline

```
Empty: 338ms
Base to Base: 397ms
Derived to Base: 588ms
Derived2 to Base: 678ms
Derived4 to Base: 1141ms
Derived6 to Base: 1056ms
List<int> to IList<int>: 649ms
List<int> to ICollection<int>: 676ms
List<int> to IEnumerable<int>: 963ms
List<int> to ICollection: 1092ms
List<double> to ICollection<int>: 1302ms
Dictionary<double, double> to IDeserializationCallback: 1264ms
List<int> to IReadOnlyCollection<int>: 983ms
List<string> to IReadOnlyCollection<object>: 988ms
List<double> to IReadOnlyCollection<int>: 989ms
string[] to IReadOnlyCollection<object>: 991ms
```

=== after the change:
```
Empty: 286ms
Base to Base: 453ms
Derived to Base: 755ms
Derived2 to Base: 816ms
Derived4 to Base: 844ms
Derived6 to Base: 994ms
List<int> to IList<int>: 679ms
List<int> to ICollection<int>: 623ms
List<int> to IEnumerable<int>: 1051ms
List<int> to ICollection: 967ms
List<double> to ICollection<int>: 867ms
Dictionary<double, double> to IDeserializationCallback: 1164ms
List<int> to IReadOnlyCollection<int>: 1234ms
List<string> to IReadOnlyCollection<object>: 1166ms
List<double> to IReadOnlyCollection<int>: 1019ms
string[] to IReadOnlyCollection<object>: 1111ms
```

VSadov commented:
Note - all the measurements have some degree of noise. 
At these level we are sensitive to alignment of code blocks and jumps/labels, to branch mispredictions and the like. So sometimes results vary for seemingly no reasons.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs:
```diff
@@ -0,0 +1,476 @@
+        }
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        private static extern object JITutil_IsInstanceOfAny_NoCacheLookup(void* toTypeHnd, object obj);
```
Nit: I would use regular managed naming convention for these. `JITutil_` prefix made sense when it was C entrypoint  or written in assembly. It does not make sense not that it is a regular method embedded within a type.

VSadov replied:
I figured it would be helpful in reviews - to see where the code moved to. But I can drop the "JITUtil_" part. 

---
In reply to: [368261375](https://github.com/dotnet/runtime/pull/1068#discussion_r368261375) [](ancestors = 368261375)

VSadov replied:
Should we remove the `JIT_` in the entry points?  Those _are_ supposed to be called by JIT only.

---
In reply to: [368261476](https://github.com/dotnet/runtime/pull/1068#discussion_r368261476) [](ancestors = 368261476,368261375)

jkotas replied:
I would remove `JIT_`. The enclosing type name `CastHelpers` is descriptive enough.

`JIT_` prefix actually misleads people into thinking that this method is part of the JIT. It was particularly confusing for `JIT_Throw`. The JIT team had to explain all the time that `JIT_Throw` on the stack does not imply that the crash is caused by the bug in the JIT. We have renamed it to `IL_Throw` to eliminate the confusion.

VSadov commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs:
```diff
@@ -227,26 +227,6 @@ private static CastResult TryGet(nuint source, nuint target)
-                    //             linear/speculative compares, and without any math vectorization does not help that much.
-                    //
-
-                    //var toTypeV = Vector256.Create((nuint)toTypeHnd);
```
One interesting thing I tried here - vectorizing interface lookup via `Avx2` - by doing 4 compares at a time. It did not help enough though. Speculative CPU seems to be good at these compares. Probably does 4 at a time already.

It was an interesting try though.

VSadov commented on src/coreclr/src/vm/castcache.h:
```diff
@@ -262,20 +262,20 @@ class CastCache
 
 #if BIT64
-        TADDR hash = (((ULONGLONG)source << 32) | ((ULONGLONG)source >> 32)) ^ target;
+        UINT64 hash = (((UINT64)source << 32) | ((UINT64)source >> 32)) ^ (UINT64)target;
```
Just if you wonder. The type changes like this are mostly to make C++ implementation as similar to C# version as possible.

VSadov committed:
Reverted removal of array helpers. They are mapped to "any" for now.

VSadov committed:
Removed `JITUtil_` and `JIT_` from managed helpers

VSadov committed:
Turn `BaseMethodTable` and `InterfaceMap` into fields.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs:
```diff
@@ -0,0 +1,476 @@
+                    {
+                        if (interfaceMap[i + 0] == (nuint)toTypeHnd)
+                            goto done;
+                        if (interfaceMap[i + 1] == (nuint)toTypeHnd)
```
This overruns the buffer. Do we need to check that there is enough remaining slots?

VSadov replied:
asm version did the same. The interface map is aligned to 4 entries and padded with zeros, if needed. That is so we could check in blocks of 4 entries. 
I will add a comment.

VSadov replied:
Also added a comment about `ParentMethodTable`. There is some subtleness there too.

---
In reply to: [368308283](https://github.com/dotnet/runtime/pull/1068#discussion_r368308283) [](ancestors = 368308283)

jkotas replied:
> asm version did the same

Can you add a link where it does it? I do not see it.

> The interface map is aligned to 4 entries and padded with zeroes, if needed. 

I do not think that we have any code that does alignment and padding like this. Also, I do not think it would be worth it - interface cast performance is not that important to pay for it.

VSadov replied:
I think I got confused by comments in the asm version. 

---
In reply to: [368312453](https://github.com/dotnet/runtime/pull/1068#discussion_r368312453) [](ancestors = 368312453)

jkotas replied:
Right, that comment is wrong.

VSadov committed:
Comments about nuances of `InterfaceMap` and `ParentMethodTable`.
Also renamed `BaseMethodTable` to `ParentMethodTable` - to match native code closer.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:
```diff
@@ -300,10 +300,41 @@ internal unsafe struct MethodTable
+        [FieldOffset(0x0e)]
+        public ushort InterfaceCount;
+        [FieldOffset(ParentMethodTableOffset)]
+        // NOTE: the ParentMethodTable could be an indirect pointer. (see: enum_flag_HasIndirectParent).
```
It can also be a relative pointer. All these tweaks apply for fragile NGen only that is disabled by default. I do not think we need to worry about fragile NGen support, or adding comments like this. It will be deleted at some point.
 
If you would like keep it working for now, you can always use the Any version of the helpers for fragile NGen - just like what you have done for interfaces.

VSadov replied:
I will remove the comment. I was not sure if that is supported or not. 
It did gave me a pause to think how we should deal with indirection.

It seems we could follow the parent chain naturally here though, even if indirect case was possible. 

---
In reply to: [368312225](https://github.com/dotnet/runtime/pull/1068#discussion_r368312225) [](ancestors = 368312225)

VSadov replied:
I will remove the comment. I was not sure if that is supported or not. 
It did gave me a pause to think how we should deal with indirection.

It seems we could follow the parent chain naturally here though, even if indirect case was possible. 

---
In reply to: [368312225](https://github.com/dotnet/runtime/pull/1068#discussion_r368312225) [](ancestors = 368312225)

jkotas commented on src/coreclr/src/vm/amd64/JitHelpers_Fast.asm:
```diff
@@ -52,378 +52,6 @@ endif
-endif
-        je      Found
-        ; move to next entry in list
-        dec     r11w
```
Here you can see that we check whether there is anything remaining after checking each entry.

VSadov committed:
check exact number of interfaces (do not assume 4-items alignment)

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:
```diff
@@ -300,10 +300,37 @@ internal unsafe struct MethodTable
+        [FieldOffset(ParentMethodTableOffset)]
+        public MethodTable* ParentMethodTable;
+        [FieldOffset(InterfaceMapOffset)]
+        public nuint* InterfaceMap;
```
```suggestion
        public MethodTable** InterfaceMap;
```

(This should simplify some of the casts.)

jkotas replied:
```suggestion
        public MethodTable** InterfaceMap;
```

(This should simplify some of the casts.)

VSadov committed:
MethodTable** InterfaceMap.
Missed an attribute in one place

VSadov committed:
comment

jkotas commented on src/coreclr/src/vm/jithelpers.cpp:
```diff
@@ -2188,365 +2188,7 @@ BOOL ObjIsInstanceOf(Object* pObject, TypeHandle toTypeHnd, BOOL throwCastExcept
-HCIMPLEND
-
-NOINLINE HCIMPL2(Object*, JITutil_ChkCastAny_NoCacheLookup, CORINFO_CLASS_HANDLE type, Object* obj)
+NOINLINE HCIMPL2(Object*, ChkCastAny_NoCacheLookup, CORINFO_CLASS_HANDLE type, Object* obj)
```
```suggestion
HCIMPL2(Object*, ChkCastAny_NoCacheLookup, CORINFO_CLASS_HANDLE type, Object* obj)
```

jkotas commented on src/coreclr/src/vm/jithelpers.cpp:
```diff
@@ -2569,7 +2211,7 @@ NOINLINE HCIMPL2(Object*, JITutil_ChkCastAny_NoCacheLookup, CORINFO_CLASS_HANDLE
 HCIMPLEND
 
-NOINLINE HCIMPL2(Object*, JITutil_IsInstanceOfAny_NoCacheLookup, CORINFO_CLASS_HANDLE type, Object* obj)
+NOINLINE HCIMPL2(Object*, IsInstanceOfAny_NoCacheLookup, CORINFO_CLASS_HANDLE type, Object* obj)
```
```suggestion
HCIMPL2(Object*, IsInstanceOfAny_NoCacheLookup, CORINFO_CLASS_HANDLE type, Object* obj)
```

jkotas commented on src/coreclr/src/vm/jithelpers.cpp:
```diff
@@ -2569,7 +2211,7 @@ NOINLINE HCIMPL2(Object*, JITutil_ChkCastAny_NoCacheLookup, CORINFO_CLASS_HANDLE
 
-NOINLINE HCIMPL2(Object*, JITutil_IsInstanceOfAny_NoCacheLookup, CORINFO_CLASS_HANDLE type, Object* obj)
+NOINLINE HCIMPL2(Object*, IsInstanceOfAny_NoCacheLookup, CORINFO_CLASS_HANDLE type, Object* obj)
 {
```
It would be nice to add `HELPER_METHOD_POLL();` right before `HELPER_METHOD_FRAME_END` to these two NoCacheLookup helpers while you are on it. It will make sure that the casting helpers are 100% GC poll friendly (fixes https://github.com/dotnet/coreclr/issues/27931).

jkotas reviewed (approved):
LGTM otherwise, but I would like David and somebody from the debugger team to sign-off too.

VSadov committed:
GC poll in 2 remaining cast FCALLs

noahfalk reviewed (approved):
From the debugger side things looked fine to me.

VSadov commented:
Thanks!!

davidwrighton reviewed (approved):
This appears semantically correct. I have 1 performance concern, which is the affect on startup. We have strong goals to make startup fast and efficient, and as written this logic will trigger jitting on startup 100% of the time. I understand that you've done some performance tuning and determined that crossgen generates code which is measurably slower. I think this change is large enough you should check in as is, but also file an issue to address the startup time regression, which a measurement of how much JIT time is spent on your machine jitting these new managed helpers, so we can prioritize making R2R compilation of these helpers great, or work on other stuff instead.

jkotas commented:
> this logic will trigger jitting on startup 100% of the time

Why is it going to trigger JITing 100% of the time? These helpers should be precompiled.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs:
```diff
@@ -0,0 +1,487 @@
+        [DebuggerHidden]
+        [StackTraceHidden]
+        [DebuggerStepThrough]
+        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
```
Ok, I see - it is because of this `[MethodImpl(MethodImplOptions.AggressiveOptimization)]`. I do not think we should have AggressiveOptimization on these.

VSadov replied:
I think I mentioned this. Tiering eventually arrives to the same codegen as jitting, but there is some minor perf penalty. Probably because methods are located/aligned differently. I am not sure what else could be different. 
I am ok with removing AggressiveOptimization.

---
In reply to: [369332603](https://github.com/dotnet/runtime/pull/1068#discussion_r369332603) [](ancestors = 369332603)

VSadov committed:
Removed `AggressiveOptimization` from casting helpers.

VSadov commented:
@davidwrighton - I have entered a tracking issue for the opposite scenario - https://github.com/dotnet/runtime/issues/1994

That is basically to determine that there is indeed a throughput impact due to tiering. 
(I am not entirely convinced, could be memory layout issues, which may change after POGO)

If that is indeed the case, we should look whether something can be improved.

VSadov commented:
Removed `AggressiveOptimization` from casting helpers.

VSadov commented:
Thanks!!

benaadams commented:
Does `build --subsetCategory coreclr -configuration release` R2R the `System.Private.CoreLib.dll` (think it does as it also produces a `System.Private.CoreLib.ni.pdb`)

If so can these be added to R2R? Am seeing `ChkCastAny` as showing up as Jitting during startup:

![image](https://user-images.githubusercontent.com/1142958/73612452-d1532000-45e3-11ea-9a24-41e48e148a89.png)

jkotas commented:
I do not see ChkCastAny JITed during startup. For the record, here is the list of methods that I see getting JITed during startup (CoreLib R2R, empty `Main`):
```
System.SpanHelpers:IndexOf(byref,ushort,int):int
System.Runtime.Intrinsics.Vector128`1[UInt16][System.UInt16]:get_Count():int
Internal.Runtime.CompilerServices.Unsafe:Add(byref,long):byref
System.Runtime.Intrinsics.Vector128:CreateScalarUnsafe(ushort):System.Runtime.Intrinsics.Vector128`1[UInt16]
System.Runtime.Intrinsics.Vector128:AsUInt32(System.Runtime.Intrinsics.Vector128`1[UInt16]):System.Runtime.Intrinsics.Vector128`1[UInt32]
System.Runtime.Intrinsics.Vector128:AsUInt16(System.Runtime.Intrinsics.Vector128`1[UInt32]):System.Runtime.Intrinsics.Vector128`1[UInt16]
System.Runtime.Intrinsics.Vector256`1[UInt16][System.UInt16]:get_Count():int
System.Text.Unicode.Utf16Utility:GetPointerToFirstInvalidChar(long,int,byref,byref):long
System.Runtime.Intrinsics.Vector128:CreateScalarUnsafe(short):System.Runtime.Intrinsics.Vector128`1[Int16]
System.Runtime.Intrinsics.Vector128:AsInt32(System.Runtime.Intrinsics.Vector128`1[Int16]):System.Runtime.Intrinsics.Vector128`1[Int32]
System.Runtime.Intrinsics.Vector128:AsInt16(System.Runtime.Intrinsics.Vector128`1[Int32]):System.Runtime.Intrinsics.Vector128`1[Int16]
System.Runtime.Intrinsics.X86.Sse41:get_IsSupported():bool
System.Runtime.Intrinsics.Vector128:AsUInt16(System.Runtime.Intrinsics.Vector128`1[Int16]):System.Runtime.Intrinsics.Vector128`1[UInt16]
System.Runtime.Intrinsics.Vector128:AsByte(System.Runtime.Intrinsics.Vector128`1[UInt16]):System.Runtime.Intrinsics.Vector128`1[Byte]
System.Text.ASCIIUtility:NarrowUtf16ToAscii(long,long,long):long
X64:get_IsSupported():bool
X64:ParallelBitExtract(long,long):long
```

What is the list of methods that you see JITed during startup?

benaadams commented:
> What is the list of methods that you see JITed during startup?

Maybe is the Profiler causing recompilation then?

![image](https://user-images.githubusercontent.com/1142958/73626590-56b8ed80-4641-11ea-9507-2fcb04a2a8ef.png)

xiangzhai commented:
:mips-interest
